packages/wrangler/src/inspect.ts

```diff
@@ -0,0 +1,526 @@
+            // eslint-disable-next-line react-hooks/exhaustive-deps
+            id: messageCounterRef.current++,
+            params: {},
+          })
+        );
+      });
+      // Finally, we'll close the websocket
+      close();
+      // And we'll clear `remoteWebsocket`
+      setRemoteWebSocket(undefined);
+    };
+  }, [
+    props.inspectorUrl,
+    retryRemoteWebSocketConnectionSigil,
+    props.logToTerminal,
+  ]);
+
+  // This effect tracks the state changes _between_ the local
+  // and remote websockets, and handles how messages flow between them.
+  useEffect(() => {
+    /**
+     * This event listener is used for buffering messages from
+     * the remote websocket, and flushing them
+     * when the local websocket connects.
+     */
+    function remoteMessageBufferListener(event: MessageEvent) {
+      if (typeof event.data === "string") {
+        if (!localWebSocket) {
+          messageBufferRef.current.push(event.data);
+          // TODO: maybe we should have a max limit on this?
+          // if so, we should be careful when removing messages
+          // from the front, because they could be critical for
+          // devtools (like execution context creation, etc)
+        }
+      } else {
+        // We should never get here, but who know is 2022...
+        console.error("unrecognised devtools event:", event);
+      }
+    }
+
+    if (remoteWebSocket && !localWebSocket) {
+      // The local websocket hasn't connected yet, so we'll
+      // buffer messages until it does.
+      remoteWebSocket.addEventListener("message", remoteMessageBufferListener);
+    }
+
+    /** Send a message from the local websocket to the remote websocket */
+    function localMessageListener(event: MessageEvent) {
+      try {
+        remoteWebSocket.send(event.data);
+      } catch (e) {
+        if (e.message !== "WebSocket is not open: readyState 0 (CONNECTING)") {
+          /**
+           * ^ this just means we haven't opened a websocket yet
+           * usually happens until there's at least one request
+           * which is weird, because we may miss something that
+           * happens on the first request. Maybe we should buffer
+           * these messages too?
+           */
+          console.error(e);
+        }
+      }
+    }
+
+    /** Send a message from the local websocket to the remote websocket */
+    function remoteMessageListener(event: MessageEvent) {
+      localWebSocket.send(event.data);
+    }
+
+    if (localWebSocket && remoteWebSocket) {
+      // Both the remote and local websockets are connected, so let's
+      // start sending messages between them.
+      localWebSocket.addEventListener("message", localMessageListener);
+      remoteWebSocket.addEventListener("message", remoteMessageListener);
+
+      // Also, let's flush any buffered messages
+      messageBufferRef.current.forEach((evt) => {
+        localWebSocket.send(evt);
+      });
```

Comment:
And if we renamed `remoteMessageListener` to `sendMessageToLocal` then it would have been more obvious to reuse that here:

```suggestion
      messageBufferRef.current.forEach(sendMessageToLocal);
```